socket.go

```diff
@@ -561,123 +613,255 @@ func (socket *mongoSocket) Query(ops ...interface{}) (err error) {
 	return err
 }
 
-func fill(r net.Conn, b []byte) error {
-	l := len(b)
-	n, err := r.Read(b)
-	for n != l && err == nil {
-		var ni int
-		ni, err = r.Read(b[n:])
-		n += ni
+// sendMessage send data to the database using the OP_MSG wire protocol
+// introduced in MongoDB 3.6 (require maxWireVersion >= 6)
+func (socket *mongoSocket) sendMessage(op *msgOp) (writeCmdResult, error) {
+	var wr writeCmdResult
+	var err error
+
+	buf := bytesBufferPool.Get().([]byte)
+
+	buf = addHeader(buf, dbMessage)
+	buf = addInt32(buf, int32(op.flags))
+
+	for _, section := range op.sections {
+		buf, err = addSection(buf, section)
+		if err != nil {
+			return wr, err
+		}
 	}
-	return err
+
+	if len(buf) > socket.ServerInfo().MaxMessageSizeBytes {
+		return wr, fmt.Errorf("message length to long, should be < %v, but was %v", socket.ServerInfo().MaxMessageSizeBytes, len(buf))
+	}
+	// set the total message size
+	setInt32(buf, 0, int32(len(buf)))
+
+	var wait sync.Mutex
+	var reply msgOp
+	var responseError error
+	var wcr writeCmdResult
+	// if no response expected, ie op.flags&opMsgFlagMoreToCome == 1,
+	// request should have id 0
+	var requestID uint32
+	// if moreToCome flag is set, we don't want to know the outcome of the message.
+	// There is no response to a request where moreToCome has been set.
+	expectReply := (op.flags & opMsgFlagMoreToCome) == 0
+
+	socket.Lock()
+	if socket.dead != nil {
+		dead := socket.dead
+		socket.Unlock()
+		debugf("Socket %p to %s: failing query, already closed: %s", socket, socket.addr, socket.dead.Error())
+		return wr, dead
+	}
+	if expectReply {
+		// Reserve id 0 for requests which should have no responses.
+	again:
+		requestID = socket.nextRequestId + 1
+		socket.nextRequestId++
+		if requestID == 0 {
+			goto again
+		}
+		wait.Lock()
+		socket.opMsgReplyFuncs[requestID] = func(msg *msgOp, err error) {
+			reply = *msg
+			responseError = err
+			wait.Unlock()
+		}
+	}
+	socket.Unlock()
+
+	setInt32(buf, 4, int32(requestID))
+	stats.sentOps(1)
+
+	socket.updateDeadline(writeDeadline)
+	_, err = socket.conn.Write(buf)
+
+	// TODO optimize this
```

Comment:
`bytesBufferPool` is used elsewhere so you'd be surprised by what's in the buffer!

The `append()` calls on a pre-allocated slice are fast enough to not worry about  :)

Reply:
Yes, it was intended to use the same pool for old and new wire protocol. In MongoDB 3.6, some calls use the new `OP_MSG` wire protocol, some other don't. The pool only contains empty slice of bytes, so it shouldn't be an issue ?